lib/spack/spack/cmd/config.py

```diff
@@ -431,6 +444,92 @@ def config_revert(args):
         tty.msg(msg.format(cfg_file))
 
 
+def config_prefer_upstream(args):
+    """Generate a packages config based on the configuration of all upstream
+    installs."""
+
+    scope = args.scope
+    if scope is None:
+        scope = spack.config.default_modify_scope('packages')
+
+    specs = spack.store.db.query(installed=[InstallStatuses.INSTALLED])
+
+    pref_specs = []
+    for spec in specs:
+        upstream = None
+        try:
+            upstream = spec.package.installed_upstream
+        except spack.repo.UnknownNamespaceError as err:
+            tty.die(
+                "Could not find package when checking spec {0} ({1}). "
+                "This is usually due to your Spack instance not being "
+                "configured to know about the upstream's repositories."
+                .format(spec.name, err.message)
+            )
+        if (upstream and not args.local) or (not upstream and args.local):
+            pref_specs.append(spec)
```

Comment:
You can avoid the corner case of a missing repo by diffing `spack.store.db.query(...)` and `spack.store.db.query_local(...)` results to get the list of specs to prefer. That way you don't have to query the package for each one, and the code is much smaller. I think it also leads to fewer DB reads since we do a DB operation for every check of `Package.installed_upstream`.

```suggestion
    all_specs = set(spack.store.db.query(installed=True))
    local_specs = set(spack.store.db.query_local(installed=True))
    pref_specs = local_specs if args.local else all_specs - local_spec
```